### PR TITLE
Format Function definitions

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR implements formatting for function definitions and async function definitions. The formatting should be fairly complete.

## Test Plan

I added a few of my own test cases and reviewed the black differences.

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #4900** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/4900" target="_blank">Graphite</a> 
    * **PR #4921** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/4921" target="_blank">Graphite</a> 
      * **PR #4922** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/4922" target="_blank">Graphite</a> 
        * **PR #4951** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/4951" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/4951?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/other/arg.rs [hunk 1]
```diff
@@ -1,12 +1,63 @@
...
-        write!(f, [not_yet_implemented(item)])
+        let Arg {
+            range,
+            arg,
+            annotation,
+            type_comment: _,
+        } = item;
+        write!(
+            f,
+            [source_text_slice(
```
Comment (konstin) [inline crates/ruff_python_formatter/src/other/arg.rs:34]:
```suggestion
            // The name of the argument
            [source_text_slice(
```

File: crates/ruff_python_formatter/src/other/arguments.rs [hunk 2]
```diff
@@ -1,12 +1,188 @@
...
+
+        let saved_level = f.context().node_level();
+        f.context_mut().set_node_level(NodeLevel::Expression);
+
+        let format_inner = format_with(|f: &mut PyFormatter| {
+            let separator = format_with(|f| write!(f, [text(","), soft_line_break_or_space()]));
+            let mut joiner = f.join_with(separator);
+            let mut last_node: Option<AnyNodeRef> = None;
+
+            let mut defaults = std::iter::repeat(None)
```
Comment (konstin) [inline crates/ruff_python_formatter/src/other/arguments.rs:36]: strange decision on the grammar or parser side to emit the default separate from the argument

File: crates/ruff_python_formatter/src/other/arguments.rs [hunk 3]
```diff
@@ -1,12 +1,188 @@
...
+                joiner.entry(&ArgumentWithDefault {
+                    argument: positional,
+                    default,
+                });
+
+                last_node = Some(default.map_or_else(|| positional.into(), AnyNodeRef::from));
+            }
+
+            if !posonlyargs.is_empty() {
+                joiner.entry(&text("/"));
```
Comment (konstin) [inline crates/ruff_python_formatter/src/other/arguments.rs:51]: so confusing that pos only arguments are in front of the slash, but keyword arguments are after the star (your code is correct)

File: crates/ruff_python_formatter/src/other/arguments.rs [hunk 4]
```diff
@@ -1,12 +1,188 @@
...
+                    kwarg.format().with_options(ArgumentKind::Kwarg)
+                ]);
+                last_node = Some(kwarg.as_any_node_ref());
+            }
+
+            joiner.finish()?;
+
+            write!(f, [if_group_breaks(&text(","))])?;
+
+            if let Some(last_node) = last_node {
```
Comment (konstin) [inline crates/ruff_python_formatter/src/other/arguments.rs:102]:
```suggestion
            // Break group if there is a trailing comma
            if let Some(last_node) = last_node {
```

Comment (konstin) [reply crates/ruff_python_formatter/src/other/arguments.rs:102]: Does this also need to read the magic trailing comma setting?

**Review by konstin — APPROVED** [review]

File: crates/ruff_python_formatter/src/other/arguments.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/other/arguments.rs:102]: If we introduce a setting for this, then yes

File: crates/ruff_python_formatter/src/other/arg.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/other/arg.rs:34]: Good point!

File: crates/ruff_python_formatter/src/other/arguments.rs (hunk 2 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/other/arguments.rs:36]: Haha yeah. I got it wrong so many times. The most recent version of `RustPython` now tracks the `default` on the argument but I haven't had time to upgrade yet.

File: crates/ruff_python_formatter/src/other/arguments.rs (hunk 4 above)
Comment (konstin) [reply crates/ruff_python_formatter/src/other/arguments.rs:102]: i forgot this isn't merged: https://github.com/astral-sh/ruff/pull/4878/files#diff-c1f1071e384c9f14936cad67780f301345744791e34afd93cada2cf43af8dcb5R18

I haven't looked into how to pass settings, but i think it makes sense to introduce the boolean for that now

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/other/arguments.rs:102]: The way this works is that Ruff should define a `PyFormatOption` struct that implements `FormatOptions` and is passed to the `PyFormatContext`. You can then access it with `f.context().options()....`

https://github.com/charliermarsh/ruff/blob/dc7cdb04e17cd9aae7ab63976424b01bf040f8d3/crates/ruff_formatter/src/lib.rs#L257-L277

**Pull request merged**